### DISCUSSION TITLE
Ability to dump F0088H0 chip from GBA carts

### DISCUSSION THREAD

**Discussion by PsychoFox11** [sanni/cartreader#1010] [Ideas & Suggestions]
The latest update included this update for flashing:
GBA: Add support for flashing 369-in-1 repro with F0088H0 flash chip

I'll admit I'm a dumper, not a flasher, but would it be possible to add dumping support? Does it do a verification afterward that's basically a dump?

This is what Limited Run's Night Trap for GBA uses, and while I was able to dump it with GBxCart RW, OSCR doesn't seem to be able to. It could likely be flashed onto this 369-in-1 lol. Any chance it could be supported, or would it have the same fate as the 64MB GBA video carts that OSCR can't do (but GBxCart RW can)?

Here's the PCB:
![image](https://github.com/user-attachments/assets/0d60baa1-e112-42e3-a5b7-55f6c1868fdd)

**Comment by PsychoFox11** [comment]
It was mentioned that I should try reading it as the 369-in-1, which I'll try tomorrow. If that works, we might just wanna make it chip specific instead of title specific. Thanks to @PsyK0p4T for suggesting this. I didn't realize that cart (or any that large) could be read!

**Reply by sanni** [reply]
It kinda looks like they just used 369-in-1 repros for their release. :see_no_evil:

**Reply by PsychoFox11** [reply]
So the result is close, but a little off, primarily bytes read by  GBxCart RW as FF are read as other bytes here and there. I'd guess it is 99% the same based on what I looked through. But it is getting late, so I'll look into it more over the weekend hopefully. It took like 45 minutes to dump. I'll dump again from the GBxCart RW just to ensure the cart's intact, and again from this one.
If they still don't match, I'll share by findings here, as well as with LesserKuma who makes Flash GBX. I know mine matched someone else's dump on that one.
I'd be interested if the two get the same for the 369-in-1 (if that hasn't been checked).

**Reply by sanni** [reply]
Edit: Never mind, tried it, doesn't help in this case. But I get read errors with my 369 too, so must be an issue with the 369 dump function

Original reply:
````
Try replacing the readWord_GBA() function in [GBA.ino line 432](https://github.com/sanni/cartreader/blob/master/Cart_Reader/GBA.ino#L432) with:

word readWord_GBA(unsigned long myAddress) {
  // Set address/data ports to output
  DDRF = 0xFF;
  DDRK = 0xFF;
  DDRC = 0xFF;

  // Divide address by two to get word addressing
  myAddress = myAddress >> 1;

  // Output address to address pins
  PORTF = myAddress;
  PORTK = myAddress >> 8;
  PORTC = myAddress >> 16;

  // Pull CS(PH3) to LOW
  PORTH &= ~(1 << 3);

  // Pull RD(PH6) to LOW
  PORTH &= ~(1 << 6);

  // Set address/data ports to input
  DDRF = 0x0;
  DDRK = 0x0;
  // Enable internal-pull-ups
  PORTF = 0xFF;
  PORTK = 0xFF;

  __asm__("nop\n\t"
          "nop\n\t"
          "nop\n\t"
          "nop\n\t");

  word myWord = (PINK << 8) | PINF;

  // Switch RD(PH6) to HIGH
  PORTH |= (1 << 6);

  // Switch CS_ROM(PH3) to HIGH
  PORTH |= (1 << 3);

  return myWord;
}

And then look if it dumps correctly.
````

**Reply by PsychoFox11** [reply]
That gives me a different CRC32, and behavior, but doesn't match what FlashGBX gives. I've confirmed it still gives the same CRC32, and that someone else using it has gotten the same hash. Though I guess that doesn't prove it's right, but I also trust @lesserkuma to know their GBA stuff!

Here's some examples of what I get now. OSCR on top.
<img width="557" alt="image" src="https://github.com/user-attachments/assets/51c140fc-a2d0-430a-b001-826ebfa3fbba">
<img width="557" alt="image" src="https://github.com/user-attachments/assets/97fe7cd8-c535-440b-805b-15e505e0fd22">
<img width="556" alt="image" src="https://github.com/user-attachments/assets/c2df1ba2-5361-4fe4-acbf-4f2a820c7d6d">

**Reply by PsychoFox11** [reply]
I saw a commit today to slow down the dumping of the 369 cart. Would that be worth a retry here as well? Just asking cause it already took over 30 min before slowdown haha.

**Reply by sanni** [reply]
It fixed the issue I had with dumping my 369 repro. How fast does the GBxCartRW dump your cart?

**Reply by Lesserkuma** [reply]
Hi! GBxCart RW maxes out at 145 KiB/s. With the newer GBFlash device, the F0088H0 369-in-1 cart can be dumped at around 685 KiB/s. Calculating CRC32 on the GBFlash itself (without dumping to a file) runs fine at 1733 KiB/s with this cart.

Dumping is done like so:
1. Set address pins as outputs.
2. Write the address.
3. Set CS pin low.
4. Set data pins as inputs.
5. Set RD pin low.
6. Delay for about 160 nanoseconds.
7. Read the ROM data from the data pins.
8. Set RD pin high.
9. Jump to step 5 until all ROM data is read.
10. Set CS pin high.

Oh, and on the older GBFlash v1.2, I had to do an extra step between steps 3 and 4 to get a clean dump. Here I had to set the data pins as outputs and write the value `FFFF` to them before setting them as inputs and doing the RD pin low. This is not necessary on the updated GBFlash v1.3 hardware revision. I think what the hardware designer did between revisions was adjusting some resistor values, but I'm not too sure.
